crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -624,26 +622,37 @@ where
     fn pending(&self) -> ProviderResult<StateProviderBox> {
         trace!(target: "providers::blockchain", "Getting provider for pending state");
 
-        // TODO: check in memory overlay https://github.com/paradigmxyz/reth/issues/9614
+        if let Some(block) = self.canonical_in_memory_state.pending_block_num_hash() {
+            let historical = self.database.history_by_block_hash(block.hash)?;
+            let pending_provider =
+                self.canonical_in_memory_state.state_provider(block.hash, historical);
+
+            return Ok(Box::new(pending_provider));
+        }
 
         // fallback to latest state if the pending block is not available
         self.latest()
     }
 
-    fn pending_state_by_hash(&self, _block_hash: B256) -> ProviderResult<Option<StateProviderBox>> {
-        // TODO: check in memory overlay https://github.com/paradigmxyz/reth/issues/9614
+    fn pending_state_by_hash(&self, block_hash: B256) -> ProviderResult<Option<StateProviderBox>> {
+        let historical = self.database.history_by_block_hash(block_hash)?;
+        if let Some(block) = self.canonical_in_memory_state.pending_block_num_hash() {
+            if block.hash == block_hash {
+                let pending_provider =
+                    self.canonical_in_memory_state.state_provider(block_hash, historical);
 
+                return Ok(Some(Box::new(pending_provider)))
+            }
+        }
         Ok(None)
     }
 
     fn pending_with_provider(
         &self,
         bundle_state_data: Box<dyn FullExecutionDataProvider>,
     ) -> ProviderResult<StateProviderBox> {
-        let canonical_fork = bundle_state_data.canonical_fork();
-        trace!(target: "providers::blockchain", ?canonical_fork, "Returning post state provider");
+        let state_provider = self.pending()?;
```

Comment:
is it ok to use the pending state provider here?

Reply:
I believe so, I actually think with the new design, this function os obsolete, need to take a closer look, for now this is okay